sematic/db/migrations/20221214142609_plugin_settings_file.py

```diff
@@ -0,0 +1,100 @@
+# Standard Library
```

Comment:
Now that we are storing a schema version in the settings file itself, we can start doing schema file migrations without using DB migrations (which is pretty hacky IMHO). Upon settings load, we can check the schema version in the config file against the current code version's schema file version. Then we can know what (if any) settings file upgrades are required.

In the code we can just keep a module that has a mapping of each schema version to functions that upgrade from the prior version to that version. We execute all the upgrade functions between the one in the file itself and the one in the code.

I guess it's fine if we start doing it this way *after* this PR since prior to this PR we don't have a reliable way to know which upgrades will need to be executed.

Reply:
If I understand correctly, you're proposing keeping the DB migration and config file migration scripts and upgrade information separate.

I would prefer to keep them in the same place in order to minimize the risk of human error, and to have a centralized record of which migrations were applied, and which not.

We could start considering the migration scripts as "external resources migration scripts" instead of just "DB migration scripts".

Reply:
Yes, that is my argument as well. Keep all migrations (DB, settings, data) in the same directory with the same tooling (migration script), and simply move the `migrations` directory one level up.

Reply:
I would argue that migrating config files and migrating dbs are fundamentally different things, because the former is per-client and the latter is per-server. Imagine this:
- two users are sharing a metadata server, even for local development
- somebody upgrades the server. This records in the server that the latest config file upgrades have happened
- another user upgrades their client. Their config file doesn't get auto-upgraded because the DB says config file upgrades migration has happened

If each person is using their own sqlite, this doesn't happen. Which is the only reason we get away with it 😆 . But this is more a coincidental success than a proper synergy of mechanisms.

Now, this isn't to say you can't have the code for config migrations handled by `migrate.py`, but it shouldn't use the same means to track which migrations have been done/need to be done.

I won't block this PR on it, because we've already done this once and so we can probably delay doing it properly. But I think we shouldn't make this a pattern.